src/cycle.rs

```diff
@@ -86,12 +86,19 @@ pub enum CycleRecoveryStrategy {
 /// A "cycle head" is the query at which we encounter a cycle; that is, if A -> B -> C -> A, then A
 /// would be the cycle head. It returns an "initial value" when the cycle is encountered (if
 /// fixpoint iteration is enabled for that query), and then is responsible for re-iterating the
-/// cycle until it converges. Any provisional value generated by any query in the cycle will track
-/// the cycle head(s) (can be plural in case of nested cycles) representing the cycles it is part
-/// of. This struct tracks these cycle heads.
+/// cycle until it converges.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct CycleHead {
+    pub database_key_index: DatabaseKeyIndex,
+    pub iteration_count: u32,
+}
+
+/// Any provisional value generated by any query in a cycle will track the cycle head(s) (can be
+/// plural in case of nested cycles) representing the cycles it is part of, and the current
+/// iteration count for each cycle head. This struct tracks these cycle heads.
 #[derive(Clone, Debug, Default)]
 #[allow(clippy::box_collection)]
-pub struct CycleHeads(Option<Box<Vec<DatabaseKeyIndex>>>);
+pub struct CycleHeads(Option<Box<Vec<CycleHead>>>);
```

Comment:
Unrelated to this PR (but something I just noticed we could investigate in a separate PR), we could check if using `ThinVec` here might be better than a `Box<Vec<_>>` (I think we have that dep nowadays)